### PR TITLE
base_encoder NaN Detection

### PR DESCRIPTION
**Pull request by Morgan-Sell**
Identifies which columns in the dataframe used in the base_encoder have null values and informs the user of those columns.

closes #351

### CODE REVIEW

File: feature_engine/encoding/base_encoder.py [hunk 1]
```diff
@@ -144,15 +144,24 @@ def transform(self, X: pd.DataFrame) -> pd.DataFrame:
...
+            # obtain the name(s) of the columns have null values
+            nan_columns = X.columns[X.isnull().any()].tolist()
+            if len(nan_columns) > 1:
+                nan_columns_str = ", ".join(nan_columns)
+            else:
+                nan_columns_str = nan_columns[0]
+
             warnings.warn(
-                "NaN values were introduced in the returned dataframe by the encoder."
+                f"NaN values were detected in the feature(s) titled {nan_columns_str}."
```
Comment (solegalli) [inline feature_engine/encoding/base_encoder.py:155]: Could we re-phrase the error sentence to this:

f"NaN values were introduced in the feature(s) {nan_columns_str} during the encoding."

**Review by solegalli — COMMENTED** [review]
Hi @Morgan-Sell 

Well done with the code changes. This is looking really good.

The tests that are failing are the code style tests. This means that something in the new code does not abide by PEP8.

To check where the error can be you can run:
`flake8 feature_engine/encoding/base_encoder.py`

That will point you to the line of code that does not follow the rules.

Probably a line of code longer than 88 characters.

File: feature_engine/encoding/base_encoder.py (hunk 1 above)
Comment (Morgan-Sell) [reply feature_engine/encoding/base_encoder.py]: I revised the warning sentence to what you suggested and fixed the formatting errors. flake8 is great!

Once a PR is in place, do I have to submit new PRs after I push edits to Git? Or, are the pushes automatically sent to the PR?

It seems that pushes are automatically forwarded to the PR.

**Comment by Morgan-Sell** [comment]
To test a warning, I suspect I use an `assert` function to determine whether the warning returns the expected string. In this case, the expected string is the warning message w/ the columns that contain NaN values.  Is this correct?

Do I create a sample dataframe that is used in the unit test? If so, where do I create the dataframe? Do I create in the `test_count_frequency_encoder.py`, the test_encoding `__init__.py` or somewhere else?

File: feature_engine/encoding/__init__.py [hunk 2]
```diff
@@ -2,6 +2,7 @@
 The module encoding includes classes to transform categorical variables into numerical.
 """
```
Comment (solegalli) [inline feature_engine/encoding/__init__.py:3]: We shouldn't change this file unless we add a new transformer. 

The BaseCategoricalTransformer is not intended for users to use. Could you please return this file to the original one?

File: feature_engine/encoding/base_encoder.py [hunk 3]
```diff
@@ -144,8 +144,16 @@ def transform(self, X: pd.DataFrame) -> pd.DataFrame:
...
+            # obtain the name(s) of the columns have null values
+            nan_columns = X.columns[X.isnull().any()].tolist()
+            if len(nan_columns) > 1:
+                nan_columns_str = ", ".join(nan_columns)
+            else:
+                nan_columns_str = nan_columns[0]
+
             warnings.warn(
-                "NaN values were introduced in the returned dataframe by the encoder."
+                f"NaN values were introduced in the feature(s) titled "
```
Comment (solegalli) [inline feature_engine/encoding/base_encoder.py:155]: Could we re-phrase this error sentence to this:

f"NaN values were introduced in the feature(s) {nan_columns_str} during the encoding."

In other words could you please remove "titled"?

File: tests/test_encoding/test_count_frequency_encoder.py [hunk 4]
```diff
@@ -2,7 +2,7 @@
 import pytest
 from sklearn.exceptions import NotFittedError
 
-from feature_engine.encoding import CountFrequencyEncoder
+from feature_engine.encoding import CountFrequencyEncoder, BaseCategoricalTransformer
```
Comment (solegalli) [inline tests/test_encoding/test_count_frequency_encoder.py:5]: after you remove BaseCategoricalTransformer from the init, the way to import it is:

from feature_engine.encoding.base_encoder import CountFrequencyEncoder, BaseCategoricalTransformer

**Comment by solegalli** [comment]
> 

Hi @Morgan-Sell 

Maybe this link helps: https://stackoverflow.com/questions/33920322/how-do-i-test-exceptions-and-errors-using-pytest

**Review by solegalli — COMMENTED** [review]
Hi @Morgan-Sell 

thank you for the changes. We should not modify the init file though. 

Would you mind having a look at my comments and see if you can revert to the original file?

thank you!

File: feature_engine/encoding/base_encoder.py (hunk 3 above)
Comment (Morgan-Sell) [reply feature_engine/encoding/base_encoder.py]: done.

File: tests/test_encoding/test_count_frequency_encoder.py (hunk 4 above)
Comment (Morgan-Sell) [reply tests/test_encoding/test_count_frequency_encoder.py:5]: How come `CountFrequencyEncoder` and other encoders classes are included in the init file and not `BaseCategoricalTransformer`?

Also, I updated the code.

File: feature_engine/encoding/__init__.py (hunk 2 above)
Comment (Morgan-Sell) [reply feature_engine/encoding/__init__.py]: Removed BaseCategorialTransformer from init file.

**Comment by Morgan-Sell** [comment]
Hi @solegalli,

Thanks for sharing the Stackoverflow article! I implemented the `pytest.raises()` and `assert` statement for `test_base_categorical_transformer_detect_nan()`.

Is it appropriate to add a `UserWarning` label to the warning text? It seems that `pytest.raises()` requires a label for the warning.

On the topic of project mgmt, is it ok for me to select "Resolve conversation" after I implement your comments? Or, are you, as the reviewer, the only person who should select "Resolve conversation"?

File: tests/test_encoding/test_count_frequency_encoder.py (hunk 4 above)
Comment (solegalli) [reply tests/test_encoding/test_count_frequency_encoder.py]: The BaseCategoricalEncoder is an internal class, not meant to be used by Feature-engine users, thus, it is not in the init.

**Comment by solegalli** [comment]
You are welcome to resolve conversations :)

**Comment by solegalli** [comment]
Hi @Morgan-Sell 

I made a [PR](https://github.com/Morgan-Sell/feature_engine/pull/1) to your repo with the necessary changes.

Could you please have a look and merge?

Thank you!

**Comment by solegalli** [comment]
Note: the decision tree encoder does not raise the warning, because the transformer raises an error when the decision tree discretizer introduces nan. 

We need to see if we are stuck here, or we can add a more informative error for the users. Probably becomes clearer after working on issue #295

**Comment by solegalli** [comment]
Hi @Morgan-Sell 

Thinking it further, we will need the encoders to be able to raise an error as well as a warning, when introducing NaN.

Do you think you could do the following:

- [ ] revise and make the PR I made to your repo
- [ ] add an init param in the encoders: rare_labels='ignore'
- [ ] extend the tests to accommodate the new functionality

If rare_labels is 'ignore', then the encoders will raise a warning as they do now. But when rare_labels is 'raise', instead of a warning, they should raise an error.

The init param needs to be added to the CountFrequencyEncoder, WoEEncoder, PRatioEncoder, MeanEncoder and OrdinalEncoder. The functionality needs to be extended in the BaseCategoricalEncoder. The tests to be extended are the ones we modified in this PR, to test for a warning or for the error.

Let me know.

Cheers

**Comment by Morgan-Sell** [comment]
Hola @solegalli,

Yes, I can work on those revisions. I've already started working on issue # [295](https://github.com/feature-engine/feature_engine/issues/295). May I complete that issue before returning to this one?

Also, I noticed that the code failed one of the tests. I see the error states `ERROR:   typechecks: commands failed`. I'm unclear as to why.

**Comment by solegalli** [comment]
Hi @Morgan-Sell ,

You can work in whichever order you prefer. 

Having said this, this PR is almost ready, and it has the logic that you need to implement in the discretizer. So If you finished this one first, you would already know what you have to do in the other PR.

But up to you. I don't really mind.

Regarding the typecheck error, it comes with the new release of mypy and we fixed it in a different PR #333

**Comment by Morgan-Sell** [comment]
Hi @solegalli,

In that case, I'll finish this task first. ;)

After merging your PR, shouldn't I see your edits in my codebase (in PyCharm)? I don't see those revisions when I have the `encoder-nan-detection` branch checked out.

I tried `git pull` to pull in your changes. However, I received the following message, **"There is no tracking information for the current branch. Please specify which branch you want to merge with."**

**Comment by solegalli** [comment]
Hi @Morgan-Sell 

Did you try searching for the error in a search engine or stackoverflow? That will allow you to troubleshoot much faster than waiting for me to reply. Also, I don't always have the right answer ;)

In this case, most likely you need to check out your branch locally:
`git checkout mybranch`

and then pull the changes you have in that branch remotely
`git pull origin mybranch`

Hope that helps
Cheers

**Comment by Morgan-Sell** [comment]
@solegalli, gracias! I don't know why I just didn't Google the error message. Google/Stackoverflow are great friends. Que boludo!

I started implementing the code to raise the error if `self.rare_labels == "raise"`. I only revised the `CountFrequencyEncoder()` test b/c I thought it would be best to approve/disapprove of the code prior to replicating the code for the other four encoding classes.

File: feature_engine/encoding/count_frequency.py [hunk 5]
```diff
@@ -97,6 +97,7 @@ def __init__(
         encoding_method: str = "count",
         variables: Union[None, int, str, List[Union[str, int]]] = None,
         ignore_format: bool = False,
+        rare_labels: str = "ignore"
```
Comment (solegalli) [inline feature_engine/encoding/count_frequency.py:100]: we need to add this parameter in the docstrings. After line 56. Could you add that?

For inspiration, look at the doscrtings of the param "errors" in pandas to datetime: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html

**Review by solegalli — COMMENTED** [review]
Hi @Morgan-Sell, 

This is looking really good. Great job, well done. 

The new functionality is ready. We need 3 tiny extras:

- [ ] Could you extend the docstrings with the new parameter?
- [ ] Could you add add a check to ensure that the users can pass only ignore or raise as values and not anything else?
- [ ] Could you add 1 test in each class to ensure that if the user passes something else in the param rare_labels, say "hello", the transformer raises an error?

In connection to that, I noticed that 3 of the parameters in the init, are shared by all transformers. Then, we are repeating the same code, over and over. To avoid boiler plate (same chunk of code over and over), we could pass the functionality of the init to the base_categorical class.

So in short, could you add an __init__ in the BaseCategorical class with the parameters `variables, ignore_format` and `rare_labels`? And migrate the functionality (check for input values and value assignment) to the base categorical class?

Then, we would need to adapt the encoders init to inherit those 3 from the base categorical, and just check and assign any new parameters if any.

For inspiration, see the WinzorizerBase [here](https://github.com/feature-engine/feature_engine/blob/0be7a33462190e3571cb2ebc90a4736edfdc0135/feature_engine/outliers/base_outlier.py#L100), and how it is updated in the init of the winzoriser [here](https://github.com/feature-engine/feature_engine/blob/0be7a33462190e3571cb2ebc90a4736edfdc0135/feature_engine/outliers/winsorizer.py#L145). 

Let me know what you think. 

Finally, apologies for the delay in the review, I was traveling to Buenos Aires :p, and it takes quite a long time to get here :p

**Comment by Morgan-Sell** [comment]
@solegalli, que bueno! da mis saludos a BsAs y disfruta un vaso de Malbec para mi. O mas, si prefieres ;)

I moved the init params - `self.variables` and `self.ignore_format` - and corresponding docstrings to the parent class BaseCategoricalTransformer. I made the following revisions to all of the child Encoding classes docstrings (hopefully they follow the appropriate standards):

- Deleted the mentioning of `ignore_format` in the docstring.
- Deleted `variables` and `ignore_format` from the "Parameters" section.
- Added a note to reference BaseCategoricalTransformer to read the definition of `variables` and `ignore_format`.

With regards to `self.rare_labels`, the parameter does **not** exist in DecisionTreeEncoder, OneHotEncoder, and RareLabelEncoder. Should we still include `rare_labels` in the parent class BaseCategoricalTransformer? 

I have not made revisions to `rare_labels`.

And, yes, I will work on the other tasks ;)

**Comment by solegalli** [comment]
> 

Hi @Morgan-Sell 

Good call about the new parameter not being present in OHE, tree discretizer and rare labels!

In fact, for just 2 parameters I think it is not worth it to migrate them to the baseclass. 

Could we then undo the last 3 commits, and add the  param value check on each relevant transformer and its respective docstring?

I think you probably need to check out the relevant commit, make the changes, and if push does not do it, then force push.

Sorry, my bad :/

I apologize.

btw, I don't like red wine. Is white ok?

**Comment by solegalli** [comment]
Hi @Morgan-Sell 

Actually, this refactoring wasn't so straightforward, so I gave it a go and sent a [PR to your repo](https://github.com/Morgan-Sell/feature_engine/pull/2)

pls have a look and merge if you are happy

Thank you!

**Comment by Morgan-Sell** [comment]
@solegalli, I merged the other PR. Thanks for the help!

Are all the edits that you requested now complete?

I'm unclear on the status of a task from one of your prior posts - "Could you add 1 test in each class to ensure that if the user passes something else in the param rare_labels, say "hello", the transformer raises an error?"

Is this task addressed by the following code?
```

if rare_labels not in ["raise", "ignore"]:
     raise ValueError(
         "rare_labels takes only values 'raise' and 'ignore ."
         f"Got {rare_labels} instead."
     )
```

Given that the `rare_labels` check is in the `__init__` method, do we still need to create a `rare_labels` check in the `test_error_if_input_df_contains_categories_not_present_in_fit_df` function?


Y, si vino blanco esta bien! En realidad, malbec es el vino rojo unico que bebo. Y, come un alfajor tambien!

**Comment by Morgan-Sell** [comment]
Hi @solegalli,

I was reviewing the code and I noticed an inconsistency.

The check/warning for the NaN values is in the `transform()` method in the BaseCategoricalTransformer class.

However, the test for NaN values in the `transform()` method is in the `test_error_if_input_df_contains_categories_not_present_in_fit_df` fcn on the **test_count_frequency_encoder.py** file.

Should the code below be moved to the `test_transform_raises_error_if_df_contains_na` fcn (test case 4)?
```
msg = "During the encoding, NaN values were introduced in the feature(s) var_A."
    encoder = CountFrequencyEncoder()

    if encoder.rare_labels == "ignore":
        with pytest.warns(UserWarning) as record:
            encoder.fit(df_enc)
            encoder.transform(df_enc_rare)

        # check that only one warning was raised
        assert len(record) == 1
        # check that the message matches
        assert record[0].message.args[0] == msg

    elif encoder.rare_labels == "raise":
        with pytest.raises(ValueError) as record:
            encoder.fit(df_enc)
            encoder.transform(df_enc_rare)

        # check that only one error was raised
        assert len(record) == 1
        # check that the error message matches
        assert record[0].message.args[0] == msg
```

`df_enc_rare` should be replaced w/ `df_enc_na`.

Meanwhile, the `test_error_if_input_df_contains_categories_not_present_in_fit_df` fcn should have code similar to the following?

```
        with pytest.raises(ValueError) as record:
            encoder = CountFrequencyEncoder()
            encoder.fit(df_enc)
            encoder.transform(df_enc_rare)
```

File: tests/test_encoding/test_count_frequency_encoder.py [hunk 6]
```diff
@@ -122,10 +122,28 @@ def test_error_if_input_df_contains_categories_not_present_in_fit_df(
 ):
     # test case 3: when dataset to be transformed contains categories not present in
     # training dataset
-    with pytest.warns(UserWarning):
-        encoder = CountFrequencyEncoder()
-        encoder.fit(df_enc)
-        encoder.transform(df_enc_rare)
+    msg = "During the encoding, NaN values were introduced in the feature(s) var_A."
```
Comment (solegalli) [inline tests/test_encoding/test_count_frequency_encoder.py:125]: this test is only testing the default parameter `rare_labels='ignore`'. We never enter rare_labels='raise'. We need to modify the test slightly.

File: tests/test_encoding/test_count_frequency_encoder.py [hunk 7]
```diff
@@ -122,10 +122,28 @@ def test_error_if_input_df_contains_categories_not_present_in_fit_df(
...
+    encoder = CountFrequencyEncoder()
```
Comment (solegalli) [inline tests/test_encoding/test_count_frequency_encoder.py:126]: replace this line by:
`encoder = CountFrequencyEncoder(rare_labels-'ignore)`

it is the default parameter, so in practice, we do not need to specify it, but since it is the parameter we are testing, it is good to specifically put it in, for readability.

File: tests/test_encoding/test_count_frequency_encoder.py [hunk 8]
```diff
@@ -122,10 +122,28 @@ def test_error_if_input_df_contains_categories_not_present_in_fit_df(
...
+
+    if encoder.rare_labels == "ignore":
```
Comment (solegalli) [inline tests/test_encoding/test_count_frequency_encoder.py:128]: remove the if, we do not need it anymore.

File: tests/test_encoding/test_count_frequency_encoder.py [hunk 9]
```diff
@@ -122,10 +122,28 @@ def test_error_if_input_df_contains_categories_not_present_in_fit_df(
...
+        with pytest.warns(UserWarning) as record:
+            encoder.fit(df_enc)
+            encoder.transform(df_enc_rare)
+
+        # check that only one warning was raised
+        assert len(record) == 1
+        # check that the message matches
+        assert record[0].message.args[0] == msg
+
```
Comment (solegalli) [inline tests/test_encoding/test_count_frequency_encoder.py:137]: now we need to specify the encoder again but with rare labels raise. Please add

encoder = CountFrequencyEncoder(rare_labels='raise')

File: tests/test_encoding/test_count_frequency_encoder.py [hunk 10]
```diff
@@ -122,10 +122,28 @@ def test_error_if_input_df_contains_categories_not_present_in_fit_df(
...
+    elif encoder.rare_labels == "raise":
```
Comment (solegalli) [inline tests/test_encoding/test_count_frequency_encoder.py:138]: now we can remove the if.

File: tests/test_encoding/test_mean_encoder.py [hunk 11]
```diff
@@ -126,11 +126,19 @@ def test_warning_if_transform_df_contains_categories_not_present_in_fit_df(
...
+    with pytest.warns(UserWarning) as record:
         encoder = MeanEncoder()
         encoder.fit(df_enc[["var_A", "var_B"]], df_enc["target"])
         encoder.transform(df_enc_rare[["var_A", "var_B"]])
 
+    # check that only one warning was raised
+    assert len(record) == 1
+    # check that the message matches
+    assert record[0].message.args[0] == msg
+
```
Comment (solegalli) [inline tests/test_encoding/test_mean_encoder.py:141]: we need to add something similar to lines 132-140 to test that it also raises an error when `rare_labels='raise'.`

File: tests/test_encoding/test_mean_encoder.py [hunk 12]
```diff
@@ -126,11 +126,19 @@ def test_warning_if_transform_df_contains_categories_not_present_in_fit_df(
 ):
     # test case 4: when dataset to be transformed contains categories not present
     # in training dataset
-    with pytest.warns(UserWarning):
+
+    msg = "During the encoding, NaN values were introduced in the feature(s) var_A."
+
+    with pytest.warns(UserWarning) as record:
         encoder = MeanEncoder()
```
Comment (solegalli) [inline tests/test_encoding/test_mean_encoder.py:133]: specify inside the transformer rare_labels='ignore' to make clear what we are testing

File: tests/test_encoding/test_ordinal_encoder.py [hunk 13]
```diff
@@ -67,11 +67,18 @@ def test_error_if_input_df_contains_categories_not_present_in_training_df(
 ):
     # test case 4: when dataset to be transformed contains categories not present
     # in training dataset
-    with pytest.warns(UserWarning):
+    msg = "During the encoding, NaN values were introduced in the feature(s) var_A."
+
+    with pytest.warns(UserWarning) as record:
         encoder = OrdinalEncoder(encoding_method="arbitrary")
```
Comment (solegalli) [inline tests/test_encoding/test_ordinal_encoder.py:73]: remove encoding_method, and replace by rare_labels='ignore'

File: tests/test_encoding/test_ordinal_encoder.py [hunk 14]
```diff
@@ -67,11 +67,18 @@ def test_error_if_input_df_contains_categories_not_present_in_training_df(
...
         encoder.fit(df_enc)
         encoder.transform(df_enc_rare)
 
+    # check that only one warning was raised
+    assert len(record) == 1
+    # check that the message matches
+    assert record[0].message.args[0] == msg
+
```
Comment (solegalli) [inline tests/test_encoding/test_ordinal_encoder.py:81]: add test to test when rare_labels=raise

File: tests/test_encoding/test_probability_ratio_encoder.py [hunk 15]
```diff
@@ -253,3 +253,19 @@ def test_logratio_on_numerical_variables(df_enc_numeric):
...
     with pytest.raises(ValueError):
         PRatioEncoder(encoding_method="other")
+
+
+def test_warn_if_transform_df_contains_categories_not_seen_in_fit(df_enc, df_enc_rare):
+    # test case 3: when dataset to be transformed contains categories not present
+    # in training dataset
+    msg = "During the encoding, NaN values were introduced in the feature(s) var_A."
+
+    encoder = PRatioEncoder(variables=None)
```
Comment (solegalli) [inline tests/test_encoding/test_probability_ratio_encoder.py:263]: replace variables=None by rare_labels='ignore'

File: tests/test_encoding/test_probability_ratio_encoder.py [hunk 16]
```diff
@@ -253,3 +253,19 @@ def test_logratio_on_numerical_variables(df_enc_numeric):
...
+    with pytest.warns(UserWarning) as record:
+        encoder.fit(df_enc[["var_A", "var_B"]], df_enc["target"])
+        encoder.transform(df_enc_rare[["var_A", "var_B"]])
+
+    # check that only one warning was raised
+    assert len(record) == 1
+    # check that the message matches
+    assert record[0].message.args[0] == msg
```
Comment (solegalli) [inline tests/test_encoding/test_probability_ratio_encoder.py:271]: add test for when rare_labels='raise'

File: tests/test_encoding/test_woe_encoder.py [hunk 17]
```diff
@@ -90,11 +90,18 @@ def test_error_target_is_not_passed(df_enc):
 def test_warn_if_transform_df_contains_categories_not_seen_in_fit(df_enc, df_enc_rare):
     # test case 3: when dataset to be transformed contains categories not present
     # in training dataset
+    msg = "During the encoding, NaN values were introduced in the feature(s) var_A."
+
     encoder = WoEEncoder(variables=None)
```
Comment (solegalli) [inline tests/test_encoding/test_woe_encoder.py:95]: replace variables=None by rare_labels='ignore'

File: tests/test_encoding/test_woe_encoder.py [hunk 18]
```diff
@@ -90,11 +90,18 @@ def test_error_target_is_not_passed(df_enc):
...
-    with pytest.warns(UserWarning):
+    with pytest.warns(UserWarning) as record:
         encoder.fit(df_enc[["var_A", "var_B"]], df_enc["target"])
         encoder.transform(df_enc_rare[["var_A", "var_B"]])
 
+    # check that only one warning was raised
+    assert len(record) == 1
+    # check that the message matches
+    assert record[0].message.args[0] == msg
+
```
Comment (solegalli) [inline tests/test_encoding/test_woe_encoder.py:104]: add test for when rare_labels='raise'

**Review by solegalli — COMMENTED** [review]
Hi @Morgan-Sell 

Thank you for bringing to my attention the tests and my previous comment. I I was ready to merge without those, lol. And those were important.

Yes, we need to add a test to corroborate that when the user enters a parameter that is not allowed in rare_labels, the transformer triggers an error. We do have the catch-up/safety code in the base class. The objective of the test is to corroborate / ensure, that our safety code is working as expected. So basically, we need to add that test to all transformers :/

An example test for inspiration can be [this one](https://github.com/feature-engine/feature_engine/blob/6437355aa47ed6e200a590ad0e4ecc09cb686413/tests/test_encoding/test_count_frequency_encoder.py#L115).

Also, I noticed that in our current tests, we are only testing that the transformers raise a warning when `rare_labels='ignore'`. But we are not testing that the transformer raises an error when `rare_labels='raise'`. So we need to add those in as well. I added comments in all the test files to point out where we should add the code.

Finally, since we need to add a few more things, it is a good opportunity to rebase the main branch onto your feature branch, so that the test that is failing, passes. We fixed it in another PR. So to rebase main onto your feature branch, you can find guidelines [here](https://feature-engine.readthedocs.io/en/pydata_template/contribute/contribute_code.html#keep-your-fork-up-to-date).

Would you mind adding these little bits to the PR so we can finish it off? I will greatly appreciate it :)

Thank you!

**Comment by solegalli** [comment]
> Hi @solegalli,
> 
> I was reviewing the code and I noticed an inconsistency.
> 
> The check/warning for the NaN values is in the `transform()` method in the BaseCategoricalTransformer class.
> 
> However, the test for NaN values in the `transform()` method is in the `test_error_if_input_df_contains_categories_not_present_in_fit_df` fcn on the **test_count_frequency_encoder.py** file.
> 
> Should the code below be moved to the `test_transform_raises_error_if_df_contains_na` fcn (test case 4)?
> 
> ```
> msg = "During the encoding, NaN values were introduced in the feature(s) var_A."
>     encoder = CountFrequencyEncoder()
> 
>     if encoder.rare_labels == "ignore":
>         with pytest.warns(UserWarning) as record:
>             encoder.fit(df_enc)
>             encoder.transform(df_enc_rare)
> 
>         # check that only one warning was raised
>         assert len(record) == 1
>         # check that the message matches
>         assert record[0].message.args[0] == msg
> 
>     elif encoder.rare_labels == "raise":
>         with pytest.raises(ValueError) as record:
>             encoder.fit(df_enc)
>             encoder.transform(df_enc_rare)
> 
>         # check that only one error was raised
>         assert len(record) == 1
>         # check that the error message matches
>         assert record[0].message.args[0] == msg
> ```
> 
> `df_enc_rare` should be replaced w/ `df_enc_na`.
> 
> Meanwhile, the `test_error_if_input_df_contains_categories_not_present_in_fit_df` fcn should have code similar to the following?
> 
> ```
>         with pytest.raises(ValueError) as record:
>             encoder = CountFrequencyEncoder()
>             encoder.fit(df_enc)
>             encoder.transform(df_enc_rare)
> ```

test case 3: tests the situation where the dataframe to transform contains categories that were not present in the train set. Then, when the transformer encoders those variables, it introduces NA, and here is where you changed the functionality, so you added the new tests.

test case 4: tests the situation where the dataframe to transform contains NA. That is a different story, the error will be raised before doing the encoding.

Bottom line, we do not need to re-organise the tests. This is OK. 

We just need to add a few tests missing. 

Thank you for checking and bring it this up.

**Comment by Morgan-Sell** [comment]
Hi @solegalli,

I implemented the tests that you mentioned. I'll resolve the failed tests later. I thought they were caused by the `assert` statement for the `ValueError` check when `rare_labels=raise`.

**Comment by Morgan-Sell** [comment]
Hola @solegalli,

I believe I implemented all of the suggested tests. Let me know if I overlooked something.

Hopefully, the "test string" for the `test_error_if_rare_labels_not_permitted_value()` function is OK. If you would like, I am happy to change it ;)

**Comment by solegalli** [comment]
Hi @Morgan-Sell 

This is OK to merge. Thank you so much for adding all the tests!

There is one detail though. Something went wrong in the rebase, because now all the files that were modified in other PRs appear in this PR.

We have 2 options. 

The first one is that you try to resolve the issue, so that we commit this PR with all your work, and we can close it here, end to end.

I think the following should work:

`
git checkout main
git pull -rebase upstream main
git checkout -yourfeaturebranch
git rebase main 
`

and now try

`git push -f origin yourfeaturebranch`

That worked for me.

If you still see all 59 files in the PR, then we can use the second approach, which is merging #351 which is basically your changes, but then it looks like it is my commit.

Let me  know if you would like to try the above and / or if I go ahead and merge #351.

Thank you!

**Comment by Morgan-Sell** [comment]
Thanks, @solegalli! It seemed to work.

I don't why all the other PR changes were merged into this one. 

I also changed the BaseCategorical init param from `rare_labels` to `errors`. The nomenclature for the encoders' param now matches the discretizers' param.

**Comment by solegalli** [comment]
Hi @Morgan-Sell. Glad you could remove the changes from other PRs. Well done.

We now have a lot of files that seem to come from your local environment.

Would you be able to remove those as well please?

**Comment by Morgan-Sell** [comment]
Hi @solegalli, I removed the appropriate directories/files. Lmk if there's anything else!

**Comment by solegalli** [comment]
Masterpiece! Thank you @Morgan-Sell

**Comment by Morgan-Sell** [comment]
Vamossss!!!

**Comment by Morgan-Sell** [comment]
@solegalli, should I delete the branch?

**Comment by solegalli** [comment]
in your repo you mean? yes
